### ISSUE TITLE
GEANT3 gheish.F breaks

### ISSUE THREAD

**Issue by ngeraks** [FairRootGroup/FairRoot#1072]
While running MpdRoot standard simulation I am getting random crashes in the gheish.F:147  
 The bug happens even when I am running only cave volume simulation (at first I speculated it was happening in calorimeters).
 Colleagues have confirmed the same issue.
Pop_OS 20.04, linux 5.11, gcc 9.3. 
I tried to pass -std=legacy to the fortran code, it did not make a difference.
Since gheish handles hadronic processes the pdg of the offending particle is  
usually 2112 or 2212 - neutrons and protons, but also helium4 and others.  
While investigating the issue in fairsoft/transport/geant3/giface/gheish.F  
I found the crash happens at  NETEST=IKPART(KPART)
when I print KPART before that I am getting random large negative numbers e.g. -1935038898, -648521280, -2129293374
This bug is present in both older versions of fairsoft and the apr21.
The typical error looks like so:
```
# 6  gheish () at /opt/buildFS/Source/geant3/giface/gheish.F:147
# 7  0x00007fe32a743bab in g3thadr () at /opt/buildFS/Source/geant3/gtrak/gthadr.F:501
# 8  0x00007fe32a74dd5f in g3track () at /opt/buildFS/Source/geant3/gtrak/gtrack.F:254
# 9  0x00007fe32a86d51a in gutrak_ () at /opt/buildFS/Source/geant3/TGeant3/TGeant3gu.cxx:582
# 10 0x00007fe32a74e97e in gtreveroot () at /opt/buildFS/Source/geant3/gtrak/gtreveroot.F:192
# 11 0x00007fe32a659b5a in g3trig () at /opt/buildFS/Source/geant3/gbase/gtrig.F:62
# 12 0x00007fe32a864fd8 in TGeant3::Gtrig (this=0x55c99612d6c0) at /opt/buildFS/Source/geant3/TGeant3/TGeant3.cxx:3358
```

**Comment by dennisklein** [comment]
moved issue to FairRoot (assuming MpdRoot uses FairRoot to call Geant3. If FairRoot is not involved, we should report this at https://github.com/vmc-project/geant3).

@ngeraks Which FairRoot version are you using?
@ngeraks Can you provide a minimal example that triggers the crash or, if that is not feasible, provide info on how to invoke MpdRoot (which macro/executable with which parameters?) in a way that is likely to trigger this crash?

**Comment by ngeraks** [comment]
FairRoot is involved, but I do not think it is the cause. FairMCApplication is calling the GEANT3 functions.
As for minimal reproduction... I am not sure other than copying our current [installation](http://mpd.jinr.ru/howto-install/), maybe you can achieve the same result only with FairRoot/templates/project_root_containers/macro/run_sim.C(10000, "TGeant3", false) and commenting pipe, magnet, newdet, so you will be faster, maybe try to generate A LOT of protons and neutrons with the box generator. I have not tried it.  
If you would like to try installing MpdRoot and replicate my running error example, you may test by runing MpdRoot/macro/mpd/runMC.C for 10000 events, GEANT3 with HADGEN generator, and also commenting out all detectors in geometry_stage1.C  
Currently my speculation is that this is somehow happening due to compiler optimizations and the KPART variable is somehow not initialized properly, hence the random large  negative numbers.

**Comment by ngeraks** [comment]
I have reinstall FS jun19_patches and FR v18.2_patches and mpdroot all of them in debug modes.
Here is the full stack, slightly more explicit
```
===========================================================
There was a crash (kSigSegmentationViolation).
This is the entire stack trace of all threads:
===========================================================
# 0  0x00007f82d509fdba in wait4 () from /usr/lib/x86_64-linux-gnu/libc.so.6
# 1  0x00007f82d500f0e7 in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
# 2  0x00007f82d58957ac in TUnixSystem::Exec (this=0x556d80c28800, shellcmd=0x556d86dc76a0 "/opt/fairsoftDebug/install/share/root/etc/gdb-backtrace.sh 2471602 1>&2") at /opt/fairsoftDebug/tools/root/core/unix/src/TUnixSystem.cxx:2119
# 3  0x00007f82d58960b1 in TUnixSystem::StackTrace (this=0x556d80c28800) at /opt/fairsoftDebug/tools/root/core/unix/src/TUnixSystem.cxx:2413
# 4  0x00007f82d5899de3 in TUnixSystem::DispatchSignals (this=0x556d80c28800, sig=kSigSegmentationViolation) at /opt/fairsoftDebug/tools/root/core/unix/src/TUnixSystem.cxx:3644
# 5  0x00007f82d589133c in SigHandler (sig=kSigSegmentationViolation) at /opt/fairsoftDebug/tools/root/core/unix/src/TUnixSystem.cxx:408
# 6  0x00007f82d5899d32 in sighandler (sig=11) at /opt/fairsoftDebug/tools/root/core/unix/src/TUnixSystem.cxx:3621
# 7  0x00007f82d588967c in textinput::TerminalConfigUnix::HandleSignal (this=0x7f82d5a356e0 <textinput::TerminalConfigUnix::Get()::s>, signum=11) at /opt/fairsoftDebug/tools/root/core/textinput/src/textinput/TerminalConfigUnix.cpp:100
# 8  0x00007f82d5889386 in (anonymous namespace)::TerminalConfigUnix__handleSignal (signum=11) at /opt/fairsoftDebug/tools/root/core/textinput/src/textinput/TerminalConfigUnix.cpp:36
# 9  <signal handler called>
# 10 0x00007f82c34d081d in gheish () at /opt/fairsoftDebug/transport/geant3/giface/gheish.F:146
# 11 0x00007f82c372ddc6 in guhadr_ () at /opt/fairsoftDebug/transport/geant3/TGeant3/TGeant3gu.cxx:179
# 12 0x00007f82c370ed3a in jumpt0_ (iadr=0x7f82c4b35120 <gcjump_+32>) at /opt/fairsoftDebug/transport/geant3/minicern/jumptn.c:74
# 13 0x00007f82c353464a in g3thadr () at /opt/fairsoftDebug/transport/geant3/gtrak/gthadr.F:501
# 14 0x00007f82c3543e6e in g3track () at /opt/fairsoftDebug/transport/geant3/gtrak/gtrack.F:254
# 15 0x00007f82c372ea2a in gutrak_ () at /opt/fairsoftDebug/transport/geant3/TGeant3/TGeant3gu.cxx:561
# 16 0x00007f82c3545444 in gtreveroot () at /opt/fairsoftDebug/transport/geant3/gtrak/gtreveroot.F:168
# 17 0x00007f82c372ea60 in gutrev_ () at /opt/fairsoftDebug/transport/geant3/TGeant3/TGeant3gu.cxx:581
# 18 0x00007f82c370ed3a in jumpt0_ (iadr=0x7f82c4b35180 <gcjump_+128>) at /opt/fairsoftDebug/transport/geant3/minicern/jumptn.c:74
# 19 0x00007f82c33c10af in g3trig () at /opt/fairsoftDebug/transport/geant3/gbase/gtrig.F:62
# 20 0x00007f82c371735b in TGeant3::Gtrig (this=0x556d86619fe0) at /opt/fairsoftDebug/transport/geant3/TGeant3/TGeant3.cxx:3315
# 21 0x00007f82c371e6e3 in TGeant3::ProcessEvent (this=0x556d86619fe0) at /opt/fairsoftDebug/transport/geant3/TGeant3/TGeant3.cxx:6558
# 22 0x00007f82c371e5eb in TGeant3::ProcessRun (this=0x556d86619fe0, nevent=10000) at /opt/fairsoftDebug/transport/geant3/TGeant3/TGeant3.cxx:6525
# 23 0x00007f82c8fceec4 in FairMCApplication::RunMC (this=0x556d83098280, nofEvents=10000) at /opt/fairrootDebug/base/sim/FairMCApplication.cxx:465
# 24 0x00007f82c8fb6caa in FairRunSim::Run (this=0x556d85eb8bd0, NEvents=10000) at /opt/fairrootDebug/base/steer/FairRunSim.cxx:308
# 25 0x00007f82c95b49ab in ?? ()
# 26 0x00007f82d08101c8 in ?? () from /opt/fairsoftDebug/install/lib/root/libCling.so
# 27 0x00007f82d4773f40 in clang::RecursiveASTVisitor<clad::utils::ReferencesUpdater>::TraverseArrayInitLoopExpr(clang::ArrayInitLoopExpr*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*)
got.plt () from /opt/fairsoftDebug/install/lib/root/libCling.so
# 28 0x00007fff59112f20 in ?? ()
# 29 0x0000556d80c9c7f0 in ?? ()
# 30 0x00007f82c95eb000 in ?? ()
# 31 0x0000000000000000 in ?? ()
===========================================================
```

**Comment by dennisklein** [comment]
For anyone reading this, here is a reproducer (not a debug build): [fairroot_issue_1072_v1.sif](https://alfa-ci.gsi.de/packages/fairroot_issue_1072_v1.sif) (created with Singularity `3.8.0`)
E.g. use it like this
```console
$ wget https://alfa-ci.gsi.de/packages/fairroot_issue_1072_v1.sif
--2021-06-30 22:09:03--  https://alfa-ci.gsi.de/packages/fairroot_issue_1072_v1.sif
Resolving alfa-ci.gsi.de (alfa-ci.gsi.de)... 140.181.3.175
Connecting to alfa-ci.gsi.de (alfa-ci.gsi.de)|140.181.3.175|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2069319680 (1.9G)
Saving to: ‘fairroot_issue_1072_v1.sif’

fairroot_issue_1072_v1.sif      100%[======================================================>]   1.93G  12.2MB/s    in 2m 42s

2021-06-30 22:11:45 (12.2 MB/s) - ‘fairroot_issue_1072_v1.sif’ saved [2069319680/2069319680]

$ singularity exec --no-home -f -w fairroot_issue_1072_v1.sif bash -l
INFO:    Converting SIF file to temporary sandbox...
Singularity:~$ ls -l /{fairsoft,fairroot,mpdroot}
/fairroot:
total 0
drwxr-xr-x. 21 root root 720 Jun 30 20:30 build
drwxr-xr-x. 22 root root 720 Jun 30 20:26 source
drwxr-xr-x.  6 root root 120 Jun 30 20:30 v18.2_patches

/fairsoft:
total 0
drwxr-xr-x. 10 root root 280 Jun 30 20:03 jun19_patches

/mpdroot:
total 0
drwxr-xr-x. 46 root root 1260 Jun 30 20:40 source

Singularity:~$ cd /mpdroot/source
Singularity:~$ . build/config.sh -p
System during compilation: Ubuntu 20.04.2 LTS
                           x86_64
System now               : Ubuntu 20.04.2 LTS
                           x86_64
Singularity:~$ root -q -l -b macro/mpd/runMC.C &> log
Singularity:~$ tail -35 log


 *** Break *** segmentation violation
MpdGeneralGenerator::ReadEvent
 Generating stack trace...
 0x00007f132e3394bf in g3track_ + 0x6df from /fairsoft/jun19_patches/lib/libgeant321.so
 0x00007f132e46f0cb in gutrak_ + 0x4b from /fairsoft/jun19_patches/lib/libgeant321.so
 0x00007f132e33a36e in gtreveroot_ + 0x3fe from /fairsoft/jun19_patches/lib/libgeant321.so
 0x00007f132e234a5a in g3trig_ + 0x6a from /fairsoft/jun19_patches/lib/libgeant321.so
 0x00007f132e4666de in TGeant3::ProcessRun(int) + 0x8e from /fairsoft/jun19_patches/lib/libgeant321.so
 0x00007f13326fced0 in FairMCApplication::RunMC(int) at /fairroot/source/base/sim/FairMCApplication.cxx:468 from /fairroot/v18.2_patches/lib/libBase.so.18.2.1
 0x00007f1339c7acbb in <unknown function>
   TKey Writing 730 bytes at address 43068107 for ID= evetest.root Title=
   TKey Writing 56 bytes at address 43068837 for ID= evetest.root Title=
Info in <TTree::TTree::~TTree>: For tree mpdsim, allocation count is 379.
Info in <TFile::~TFile>: dtor called for evetest.root [55a2c73268a0]
Info in <TDirectoryFile::~TDirectoryFile>: dtor called for evetest.root
Info in <TDirectory::~TDirectory>: dtor called for evetest.root
 GTREVE_ROOT : Transporting primary track No        600
 GTREVE_ROOT : Transporting primary track No        500
 GTREVE_ROOT : Transporting primary track No        400
 GTREVE_ROOT : Transporting primary track No        300
 GTREVE_ROOT : Transporting primary track No        200
 GTREVE_ROOT : Transporting primary track No        100
 **** GTRIGI: IEVENT=    871 IDEVT=    871 Random Seeds =       4357           0
 GTREVE_ROOT : Transporting primary track No        100
 **** GTRIGI: IEVENT=    872 IDEVT=    872 Random Seeds =       4357           0
 GTREVE_ROOT : Transporting primary track No        700
 GTREVE_ROOT : Transporting primary track No        600
 GTREVE_ROOT : Transporting primary track No        500
 GTREVE_ROOT : Transporting primary track No        400
 GTREVE_ROOT : Transporting primary track No        300
 GTREVE_ROOT : Transporting primary track No        200
 GTREVE_ROOT : Transporting primary track No        100
 **** GTRIGI: IEVENT=    873 IDEVT=    873 Random Seeds =       4357           0
```

@ngeraks Will have a debugging session with it tomorrow, didn't get to it yet.

**Comment by dennisklein** [comment]
@ngeraks Do you know roughly, when this started to break (with a change of FairRoot, FairSoft, Geant3 and/or the Compiler for example?)?

**Comment by ngeraks** [comment]
@dennisklein Unfortunately not. I have been using for most of the time in the past 2-3 years Geant4. Previously I do not remember such crashes. I can confirm this is happening on my gcc10.3 and a colleague confirmed they have experienced it on our production cluster here is their stack trace:
```
#3  0x00007f895bee90dc in TUnixSystem::DispatchSignals(ESignals) () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/root/libCore.so.6.16
#4  <signal handler called>
#5  0x00007f893fbe6de2 in gheish_ () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#6  0x00007f893fc32942 in g3tneut_ () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#7  0x00007f893fc37fd7 in g3track_ () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#8  0x00007f893fd7cb76 in gutrak_ () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#9  0x00007f893fc390b7 in gtreveroot_ () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#10 0x00007f893fb2959e in g3trig_ () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#11 0x00007f893fd7504c in TGeant3::ProcessRun(int) () from /cvmfs/hybrilit.jinr.ru/sw/slc7_x86-64/FAIR_june19p2_gcc485/fairsoft/install/lib/libgeant321.so
#12 0x00007f894df083b9 in FairMCApplication::RunMC (this=0x6cc8490, nofEvents=1000) at /root/zuev/FairRoot-18.2.1/base/sim/FairMCApplication.cxx:465
```

**Comment by ngeraks** [comment]
Here is where I have traced the bug to:
```
C --- SET THE GHEISHA PARTICLE TYPE TO THE ONE OF GEANT ---
      print *, 'ipart ', IPART
      IF(IPART.GT.79) THEN
         print *, 'jpart ipart ', JPART, IPART
         IF(ISTOP.EQ.0) GOTO 9999
         JPA = LQ(JPART-IPART)
         print *, 'jpa ', JPA
         AMAS=Q(JPA+7)
         print *, 'amas ', AMAS
         NCH =Q(JPA+8)
         print *, 'nch ', NCH
         KPART=-IPART
         print *, 'kpart ', KPART
         GOTO 107
      ENDIF
      print *, 'kpart ', KPART
      NETEST=IKPART(KPART)
      print *, 'ntest ', NETEST
```
^This is code in fairsoft/transport/geant3/giface/gheish.F:137
It is enough to 
```
rm fairsoft/install/lib/libgeant321.so && ./configure.sh 
```
to rebuild the offending code.
This typically produces output like so
```
 ipart           13
 kpart           16
 ntest           13
 ipart           11
 kpart           10
 ntest           11
 ipart           14
 kpart           14
 ntest           14
 ipart           14
 kpart           14
 ntest           14
 ipart            9
 kpart            9
 ntest            9
 ipart           14
 kpart           14
 ntest           14
 ipart           14
 kpart           14
 ntest           14
 ipart           13
 kpart           16
 ntest           13
 ipart           14
 kpart           16
 ntest           13
 ipart           13
 kpart           14
 ntest           14
 ipart           14
 kpart           14
 ntest           14
 ipart           13
 kpart           14
 ntest           14
 ipart           46
 kpart           31
 ntest           46
```
But when I have the break:
```
 ipart           14
 kpart  -1967532960
```
I believe 14 corresponds to pdg 2112.

**Comment by ngeraks** [comment]
I am beginning to think this issue is only present in the HADGEN model which is being used.  
It is using the TObjArray* TGenerator::ImportParticles() function from root's TGenerator  
Crash is happening at:
```
GeneralGen: kf=1000791960, p=(0.00, 0.00, 1273409.25) GeV, x=(0.0, 0.0, 0.0) cm
 ipart           45
 kpart   1212266547
```

**Comment by dennisklein** [comment]
> You may consider closing this issue, as it seems to be an error in the way the model is passing particles.

I will defer the decision to one of our next internal group meetings to get some feedback from our Geant3 experts (which I am not ;) ).

**Comment by fuhlig1** [comment]
Hi @ngeraks, @dennisklein,

in my opinion this is really a problem of the used generator. I have never seen this error before in any of the different experiments in over 10 years. Also an energy of 1.2 PeV is probably something which is completely not covered by Geant3. ;-} 
I am not sure if Geant4 would make something useful with such a particle.
You have always to keep in mind that Geant3 is an old Fortran program and we interface it from C++. I am nearly sure that the problem arises when the data is hand over from C++ to Fortran.  In Fortran the data is normally stored in Common Blocks which in the  end is consecutive memory, so I think that it is very likely that you simply overwrite some memory with the seen effect.

@ngeraks could you please introduce some statements in your C++ code which simply filters particles with unphysical momenta and check if the crash in gheish.F is then gone.

**Comment by ngeraks** [comment]
@fuhlig1,
Yes. This is exactly the problem. I guess the model submits improper blocks in rare cases and this leads to memory problems.
This is why I suggested closing the issue.
It is user error. And I can not take responsibility for what the author intended. I submitted the issue without knowledge that this was in fact a wrong interfacing with Geant3. If one decides to support the model, this should be corrected.
